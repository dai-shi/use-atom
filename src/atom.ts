export type Atom<Value> = {
  init?: Value;
  read: (get: <V>(a: Atom<V>) => V) => Value | Promise<Value>;
}

export type WritableAtom<Value> = Atom<Value> & {
  write: (
    get: <V>(a: Atom<V>) => V,
    set: <V>(a: WritableAtom<V>, v: V) => void,
    newValue: Value
  ) => void;
}

export function atom<Value>(
  read: (get: <V>(a: Atom<V>) => V) => Value | Promise<Value>,
  write: (
    get: <V>(a: Atom<V>) => V,
    set: <V>(a: WritableAtom<V>, v: V) => void,
    newValue: Value,
  ) => void,
): WritableAtom<Value>;

export function atom<Value>(
  read: (get: <V>(a: Atom<V>) => V) => Value | Promise<Value>,
): Atom<Value>;

export function atom<Value>(
  initialValue: Value,
): WritableAtom<Value>;

export function atom<Value>(
  read: Value | ((get: <V>(a: Atom<V>) => V) => Promise<Value>),
  write?: (
    get: <V>(a: Atom<V>) => V,
    set: <V>(a: WritableAtom<V>, v: V) => void,
    newValue: Value,
  ) => void,
) {
  if (typeof read !== 'function') {
    const a: WritableAtom<Value> = {
      init: read,
      read: (get) => get(a),
      write: (_get, set, newValue) => set(a, newValue),
    };
    return a;
  }
  if (write) {
    const a: WritableAtom<Value> = {
      read: (get) => get(a),
      write: (_get, set, newValue) => set(a, newValue),
    };
    return a;
  }
  const a: Atom<Value> = {
    read: (get) => get(a),
  };
  return a;
}
