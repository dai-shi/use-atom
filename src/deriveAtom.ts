import { Atom, WritableAtom } from './createAtom';

export function deriveAtom<Value>(
  options: {
    get: (arg: {
      get: <V>(a: Atom<V>) => V;
    }) => Promise<Value>;
    set: (arg: {
      get: <V>(a: Atom<V>) => V;
      set: <V>(a: WritableAtom<V>, v: V) => void;
    }, newValue: Value) => void | Promise<void>;
  },
): WritableAtom<Value | null>;

export function deriveAtom<Value>(
  options: {
    get: (arg: {
      get: <V>(a: Atom<V>) => V;
    }) => Value;
    set: (arg: {
      get: <V>(a: Atom<V>) => V;
      set: <V>(a: WritableAtom<V>, v: V) => void;
    }, newValue: Value) => void | Promise<void>;
  },
): WritableAtom<Value>;

export function deriveAtom<Value>(
  options: {
    get: (arg: {
      get: <V>(a: Atom<V>) => V;
    }) => Promise<Value>;
  },
): Atom<Value | null>;

export function deriveAtom<Value>(
  options: {
    get: (arg: {
      get: <V>(a: Atom<V>) => V;
    }) => Value;
  },
): Atom<Value>;

export function deriveAtom<Value>(
  options: {
    get: (arg: {
      get: <V>(a: Atom<V>) => V;
    }) => Value | Promise<Value>;
    set?: (arg: {
      get: <V>(a: Atom<V>) => V;
      set: <V>(a: WritableAtom<V>, v: V) => void;
    }, newValue: Value) => void | Promise<void>;
  },
) {
  const atom = {
    ...options,
    default: null as Value | null,
  };
  const value = atom.get({
    get: (a) => a.default,
  });
  if (value instanceof Promise) {
    value.then((v) => {
      atom.default = v;
    });
  } else {
    atom.default = value;
  }
  return atom;
}
