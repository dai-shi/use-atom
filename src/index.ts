export { Provider } from './Provider';
export { atom } from './atom';
export { useAtom } from './useAtom';
export { useAtomValue } from './useAtomValue';
export { useSetAtom } from './useSetAtom';
