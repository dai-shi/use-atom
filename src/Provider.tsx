import React, {
  Dispatch,
  SetStateAction,
  useCallback,
  useState,
} from 'react';
import { createContext } from 'use-context-selector';

import { Atom, WritableAtom } from './atom';

const PROMISE_RESULT = Symbol();
const PROMISE_ERROR = Symbol();

const warningObject = new Proxy({}, {
  get() { throw new Error('Please use <Provider>'); },
  apply() { throw new Error('Please use <Provider>'); },
});

type InitAction = {
  type: 'INIT_ATOM';
  atom: Atom<unknown>;
  id: symbol;
};

type DisposeAction = {
  type: 'DISPOSE_ATOM';
  atom: Atom<unknown>;
  id: symbol;
};

type UpdateAction = {
  type: 'UPDATE_VALUE';
  atom: WritableAtom<unknown>;
  update: SetStateAction<unknown>;
};

type Action = InitAction | DisposeAction | UpdateAction;

type AtomState<Value> = {
  promise: Promise<Value> | null;
  value?: Value;
  dependents: Set<Atom<unknown> | symbol>; // symbol is id from INIT_ATOM
};

type State = Map<Atom<unknown>, AtomState<unknown>>;

const initialState: State = new Map();

const getAtomState = <Value, >(state: State, atom: Atom<Value>) => {
  const atomState = state.get(atom) as AtomState<Value> | undefined;
  if (!atomState) {
    throw new Error('atom is not initialized');
  }
  return atomState;
};

export const getAtomStateValue = <Value, >(state: State, atom: Atom<Value>) => {
  const atomState = state.get(atom) as AtomState<Value> | undefined;
  if (!atomState) {
    if ('init' in atom) return atom.init as Value;
    throw new Error('no atom init');
  }
  if (atomState.promise) {
    const {
      [PROMISE_RESULT]: result,
      [PROMISE_ERROR]: error,
    } = atomState.promise as any; // eslint-disable-line @typescript-eslint/no-explicit-any
    if (error) throw error;
    if (result) return result as Value;
    throw atomState.promise;
  }
  if ('value' in atomState) return atomState.value as Value;
  throw new Error('no atom value');
};

function appendMap<K, V>(dst: Map<K, V>, src: Map<K, V>) {
  src.forEach((v, k) => {
    dst.set(k, v);
  });
  return dst;
}

const initAtom = (
  prevState: State,
  setState: Dispatch<SetStateAction<State>>,
  atom: Atom<unknown>,
  dependent: Atom<unknown> | symbol,
) => {
  let atomState = prevState.get(atom);
  if (atomState) {
    const nextAtomState = {
      ...atomState,
      dependents: new Set(atomState.dependents).add(dependent),
    };
    return new Map().set(atom, nextAtomState);
  }
  const updateState: State = new Map();
  let isSync = true;
  const nextValue = atom.read(<V, >(a: Atom<V>) => {
    if (a !== atom) {
      if (isSync) {
        appendMap(updateState, initAtom(prevState, setState, a, atom));
      } else {
        setState((prev) => appendMap(
          new Map(prev),
          initAtom(prev, setState, a, atom),
        ));
      }
    }
    return getAtomStateValue(prevState, a);
  });
  if (nextValue instanceof Promise) {
    const promise = nextValue.then((value) => {
      setState((prev) => new Map(prev).set(atom, {
        ...getAtomState(prev, atom),
        value,
      }));
    });
    atomState = {
      promise,
      dependents: new Set(),
    };
  } else {
    atomState = {
      promise: null,
      value: nextValue,
      dependents: new Set(),
    };
  }
  atomState.dependents.add(dependent);
  updateState.set(atom, atomState);
  isSync = false;
  return updateState;
};

const disposeAtom = (
  prevState: State,
  dependent: Atom<unknown> | symbol,
) => {
  let nextState = new Map(prevState);
  const deleted: Atom<unknown>[] = [];
  nextState.forEach((atomState, atom) => {
    if (atomState.dependents.has(dependent)) {
      const nextGetDependents = new Set(atomState.dependents);
      nextGetDependents.delete(dependent);
      if (nextGetDependents.size) {
        nextState.set(atom, {
          ...atomState,
          dependents: nextGetDependents,
        });
      } else {
        nextState.delete(atom);
        deleted.push(atom);
      }
    }
  });
  nextState = deleted.reduce((p, c) => disposeAtom(p, c), nextState);
  return nextState;
};

const updateValue = (
  prevState: State,
  setState: Dispatch<SetStateAction<State>>,
  action: UpdateAction,
) => {
  const nextState = new Map(prevState);
  let isSync = true;
  const valuesToUpdate = new Map<Atom<unknown>, unknown>();
  const promises: Promise<void>[] = [];

  const getCurrAtomValue = <Value, >(atom: Atom<Value>) => {
    if (valuesToUpdate.has(atom)) {
      return valuesToUpdate.get(atom) as Value;
    }
    const atomState = nextState.get(atom) as AtomState<Value> | undefined;
    if (atomState) {
      if ('value' in atomState) return atomState.value as Value;
    }
    if ('init' in atom) {
      return atom.init as Value;
    }
    throw new Error('no atom init');
  };

  const updateDependents = (atom: Atom<unknown>) => {
    const atomState = nextState.get(atom);
    if (!atomState) return;
    atomState.dependents.forEach((dependent) => {
      if (typeof dependent === 'symbol') return;
      const v = dependent.read(<V, >(a: Atom<V>) => {
        if (a !== dependent) {
          if (isSync) {
            appendMap(nextState, initAtom(prevState, setState, a, dependent));
          } else {
            setState((prev) => appendMap(
              new Map(prev),
              initAtom(prev, setState, a, dependent),
            ));
          }
        }
        return getCurrAtomValue(a);
      });
      if (v instanceof Promise) {
        promises.push(v.then((vv) => {
          valuesToUpdate.set(dependent, vv);
        }));
      } else {
        valuesToUpdate.set(dependent, v);
      }
      updateDependents(dependent);
    });
  };

  const updateAtomValue = (atom: Atom<unknown>, value: unknown) => {
    valuesToUpdate.set(atom, value);
    updateDependents(atom);
  };

  const setValue = <Value, >(atom: WritableAtom<Value>, value: Value) => {
    const promise = atom.write(
      getCurrAtomValue,
      <V, >(a: WritableAtom<V>, v: V) => {
        if (a === atom as unknown as WritableAtom<V>) {
          updateAtomValue(atom, v);
        } else {
          setValue(a, v);
        }
      },
      value,
    );
    if (promise instanceof Promise) {
      promises.push(promise);
    }
  };

  const nextValue = typeof action.update === 'function' ? (
    action.update(getCurrAtomValue(action.atom))
  ) : (
    action.update
  );
  setValue(action.atom, nextValue);

  const hasPromises = promises.length > 0;
  const resolve = async () => {
    if (promises.length) {
      const promisesToWait = promises.splice(0, promises.length);
      await Promise.all(promisesToWait);
      await resolve();
    } else {
      setState((prev) => {
        const nextS = new Map(prev);
        valuesToUpdate.forEach((value, atom) => {
          const atomState = getAtomState(nextS, atom);
          nextS.set(atom, { ...atomState, promise: null, value });
        });
        return nextS;
      });
    }
  };
  const promise = resolve();

  if (hasPromises) {
    promise.then(() => {
      const atomState = getAtomState(nextState, action.atom);
      nextState.set(action.atom, { ...atomState, promise: null });
    });
    const atomState = getAtomState(nextState, action.atom);
    nextState.set(action.atom, { ...atomState, promise });
  }
  valuesToUpdate.forEach((value, atom) => {
    const atomState = getAtomState(nextState, atom);
    nextState.set(atom, { ...atomState, promise: null, value });
  });
  valuesToUpdate.clear();
  isSync = false;
  return nextState;
};

export const DispatchContext = createContext(warningObject as Dispatch<Action>);
export const StateContext = createContext(warningObject as State);

export const Provider: React.FC = ({ children }) => {
  const [state, setState] = useState(initialState);
  const dispatch = useCallback((action: Action) => setState((prevState) => {
    if (action.type === 'INIT_ATOM') {
      const updateState = initAtom(prevState, setState, action.atom, action.id);
      return appendMap(new Map(prevState), updateState);
    }
    if (action.type === 'DISPOSE_ATOM') {
      return disposeAtom(prevState, action.id);
    }
    if (action.type === 'UPDATE_VALUE') {
      const atomState = getAtomState(prevState, action.atom);
      if (atomState.promise) {
        const promise = atomState.promise.then(() => {
          setState((prev) => {
            const nextState = updateValue(prev, setState, action);
            return nextState;
          });
        });
        return new Map(prevState).set(action.atom, { ...atomState, promise });
      }
      const nextState = updateValue(prevState, setState, action);
      return nextState;
    }
    throw new Error('unexpected action type');
  }), []);
  return (
    <DispatchContext.Provider value={dispatch}>
      <StateContext.Provider value={state}>
        {children}
      </StateContext.Provider>
    </DispatchContext.Provider>
  );
};
