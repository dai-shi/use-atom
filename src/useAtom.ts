import {
  Dispatch,
  SetStateAction,
  useCallback,
  useEffect,
} from 'react';
import { useContext, useContextSelector } from 'use-context-selector';

import { StateContext, DispatchContext, AtomState } from './Provider';
import { Atom, WritableAtom } from './createAtom';

const isWritable = (
  atom: Atom<unknown> | WritableAtom<unknown>,
): atom is WritableAtom<unknown> => !!(atom as WritableAtom<unknown>).set;

export function useAtom<Value>(atom: WritableAtom<Value>): [Value, Dispatch<SetStateAction<Value>>]
export function useAtom<Value>(atom: Atom<Value>): [Value, never]

export function useAtom<Value>(atom: Atom<Value> | WritableAtom<Value>) {
  const dispatch = useContext(DispatchContext);
  const atomState = useContextSelector(StateContext, useCallback((state) => (
    state.get(atom) as AtomState<Value> | undefined
  ), [atom]));
  const setAtom = useCallback((update: SetStateAction<Value>) => {
    if (isWritable(atom)) {
      dispatch({ type: 'UPDATE_VALUE', atom, update });
    } else {
      throw new Error('not writable atom');
    }
  }, [atom, dispatch]);
  useEffect(() => {
    dispatch({ type: 'INIT_ATOM', atom });
    return () => {
      dispatch({ type: 'DISPOSE_ATOM', atom });
    };
  }, [dispatch, atom]);
  if (atomState && atomState.promise) {
    throw atomState.promise;
  }
  return [
    atomState === undefined ? atom.default : atomState.value,
    setAtom,
  ];
}
