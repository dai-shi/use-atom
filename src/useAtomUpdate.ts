import { SetStateAction, useCallback, useEffect } from 'react';
import { useContext } from 'use-context-selector';
import { DispatchContext } from './Provider';
import { WritableAtom } from './createAtom';

export function useAtomUpdate<Value>(atom: WritableAtom<Value>) {
  const dispatch = useContext(DispatchContext);
  const setAtom = useCallback((update: SetStateAction<Value>) => {
    dispatch({
      type: 'UPDATE_VALUE',
      atom: atom as WritableAtom<unknown>,
      update,
    });
  }, [atom, dispatch]);
  useEffect(() => {
    dispatch({ type: 'INIT_ATOM', atom });
    return () => {
      dispatch({ type: 'DISPOSE_ATOM', atom });
    };
  }, [dispatch, atom]);
  return setAtom;
}
