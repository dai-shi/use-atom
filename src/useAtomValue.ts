import { useCallback, useEffect } from 'react';
import { useContext, useContextSelector } from 'use-context-selector';
import type { Atom } from 'jotai/vanilla';
import { StateContext, DispatchContext, getAtomState } from './Provider';

export function useAtomValue<Value>(atom: Atom<Value>) {
  const dispatch = useContext(DispatchContext);
  const atomState = useContextSelector(
    StateContext,
    useCallback((state) => getAtomState(state, atom), [atom]),
  );
  useEffect(() => {
    if (!atomState.dependents) {
      dispatch({ type: 'COMMIT_ATOM', atom, atomState });
    }
  });
  useEffect(() => {
    const id = Symbol();
    dispatch({ type: 'INIT_ATOM', atom, id });
    return () => {
      dispatch({ type: 'DISPOSE_ATOM', atom, id });
    };
  }, [dispatch, atom]);
  return atomState.value;
}
