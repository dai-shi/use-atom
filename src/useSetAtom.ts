import { useCallback } from 'react';
import { useContext, useContextUpdate } from 'use-context-selector';
import type { Atom, WritableAtom } from 'jotai/vanilla';
import { StateContext, DispatchContext } from './Provider';

const isWritable = <Value, Args extends unknown[]>(
  atom: Atom<Value> | WritableAtom<Value, Args, void>,
): atom is WritableAtom<Value, Args, void> => !!(atom as WritableAtom<Value, Args, void>).write;

export function useSetAtom<Value, Args extends unknown[]>(
  atom: WritableAtom<Value, Args, void>,
) {
  const dispatch = useContext(DispatchContext);
  const updateState = useContextUpdate(StateContext);
  const setAtom = useCallback((...args: Args) => {
    if (isWritable(atom)) {
      updateState(() => {
        dispatch({
          type: 'SET_ATOM',
          atom: atom as WritableAtom<unknown, unknown[], void>,
          args,
        });
      });
    } else {
      throw new Error('not writable atom');
    }
  }, [atom, dispatch, updateState]);
  return setAtom;
}
