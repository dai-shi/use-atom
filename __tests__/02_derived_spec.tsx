import React, { useEffect, useRef, StrictMode } from 'react';
import { render, fireEvent, cleanup } from '@testing-library/react';

import { Provider, atom, useAtom } from '../src/index';

describe('derived spec', () => {
  afterEach(cleanup);

  const initialState = {
    count: 0,
    dummy: 0,
  };

  type Action = { type: 'setCount'; value: number };
  const reducer = (state = initialState, action: Action) => {
    switch (action.type) {
      case 'setCount':
        return {
          ...state,
          count: action.value,
        };
      default:
        return state;
    }
  };

  it('counter', () => {
    const globalAtom = atom(initialState);

    const countAtom = atom(
      (get) => get(globalAtom).count,
      (get, set, update: (prev: number) => number) => {
        set(globalAtom, reducer(get(globalAtom), {
          type: 'setCount',
          value: update(get(globalAtom).count),
        }));
      },
    );

    const Counter1 = () => {
      const [count1, setCount] = useAtom(countAtom);
      const increment = () => setCount((c) => c + 1);
      const commitCount = useRef(0);
      useEffect(() => {
        commitCount.current += 1;
      });
      return (
        <div>
          <span>{count1}</span>
          <button type="button" onClick={increment}>+1</button>
          <span>{commitCount.current}</span>
        </div>
      );
    };
    const App = () => (
      <StrictMode>
        <Provider>
          <Counter1 />
        </Provider>
      </StrictMode>
    );
    const { getAllByText, container } = render(<App />);
    expect(container).toMatchSnapshot();
    fireEvent.click(getAllByText('+1')[0]);
    expect(container).toMatchSnapshot();
  });
});
