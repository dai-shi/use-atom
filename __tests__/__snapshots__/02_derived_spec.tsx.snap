// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`derived spec counter 1`] = `
<div>
  <div>
    <span>
      0
    </span>
    <button
      type="button"
    >
      +1
    </button>
    <span>
      1
    </span>
  </div>
</div>
`;

exports[`derived spec counter 2`] = `
<div>
  <div>
    <span>
      1
    </span>
    <button
      type="button"
    >
      +1
    </button>
    <span>
      2
    </span>
  </div>
</div>
`;
